### PR TITLE
[FIZZ-172] Tester application

### PR DESCRIPTION
**Pull request by stephenfuqua**
You can see the TeamCity integration in action at https://intedfitools1.msdf.org/project/Experimental_SFuqua_RD_Exchange_Fizz?mode=builds#all-projects

Try running the LMS File Tester on the Google or Schoology output. I've already logged a few bugs for Schoology and will look closely at any bugs that need to be logged for Google Classroom.

### CODE REVIEW

File: src/file-tester/lms_file_tester/__main__.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import sys
+from typing import List, Union
+
+# This next line *should* have
+# `lms_file_tester.validators.directory_validation`, but the script does not run
+# from the command line if you have `lms_file_tester` there - you get a module
+# not found error. This does not make sense, and in fact mypy complains about
+# it. But pylance does not. Headscratcher. Have tried renaming file to `main.py`
+# and that did not solve it.
+from validators.directory_validation import (
```
Comment (stephenfuqua) [inline src/file-tester/lms_file_tester/__main__.py:17]: @bradbanister @lalrg please see the note above. Can you try pulling down this branch and running the script as is, and try running with the expected `from lms_file_tester.validators...` line? I'm quite perplexed about this.

To run from the `file-tester` directory: 

```
poetry run python lms_file_tester ../../docs/sample-out
```

Comment (lalrg) [reply src/file-tester/lms_file_tester/__main__.py:17]: I get the same behavoir. 
Note: with python 3.9.

Comment (stephenfuqua) [reply src/file-tester/lms_file_tester/__main__.py]: Well, I'm able to suppress the mypy error anyway.

**Comment by stephenfuqua** [comment]
@lalrg @bradbanister this is ready to review. If either of you have time to review & test today then we can do so... but if it isn't going to make it in time for sprint review tomorrow, then I want to take a few more minutes to add quickly add some TeamCity support.

**Comment by stephenfuqua** [comment]
I nearly forgot about adding docstrings to public functions. Working on that this morning.

File: .teamcity/_Self/FileTester/Project.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    buildType(fileTester.buildTypes.TestPullRequest)
+    buildType(fileTester.buildTypes.TestBranch)
+
+    params{
+        param("project.directory", "./utils/file-tester");
+        param("vcs.checkout.rules","""
+            +:.teamcity => .teamcity
+            +:%project.directory% => %project.directory%
+            +:eng => eng
+            +:src/file-utils => src/file-utils
```
Comment (stephenfuqua) [inline .teamcity/_Self/FileTester/Project.kt:24]: Had to add an extra line here to account for the fact that the File Tester application has a hard-coded dependency on the File Utils package - thus need to checkout `src/file-utils` in addition to the regular checkouts.

File: src/file-utils/lms_file_utils/directory_repository.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        Directory as string or None if the submissions directory does not exist.
+    """
+    assignments = _get_directory_for_section(
+        base_directory, section_id, f"assignment={assignment_id}"
+    )
+
+    if assignments is None:
+        return None
+
+    return os.path.join(assignments, "submissions")
```
Comment (lalrg) [inline src/file-utils/lms_file_utils/directory_repository.py:176]: I'd move strings to the top as constants and then call them when needed.
`SUBMISSIONS = "submissions"`
```suggestion
    return os.path.join(assignments, SUBMISSIONS)

```

Comment (lalrg) [reply src/file-utils/lms_file_utils/directory_repository.py:176]: Same for all strings

File: utils/file-tester/lms_file_tester/__main__.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    logging.basicConfig(
+        handlers=[
+            logging.StreamHandler(sys.stdout),
+        ],
+        format=logFormatter,
+        level=level,
+    )
+
+
+def _report(errors: List[str], happy: str):
```
Comment (lalrg) [inline utils/file-tester/lms_file_tester/__main__.py:56]: Happy? :laughing:

File: utils/file-tester/lms_file_tester/__main__.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    errors = validate_system_activities_file(input_directory)
+    _report(errors, "System Activities file is valid.")
+
+
+def _main():
+
+    if len(sys.argv) != 2:
+        logger.critical(
+            "Must pass an input directory as the only argument to the script."
```
Comment (lalrg) [inline utils/file-tester/lms_file_tester/__main__.py:134]: I think this validation should be mentioned in a readme file.

**Review by lalrg — CHANGES_REQUESTED** [review]
I just left some comments.
file-utils folder should have a readme file

File: src/file-utils/lms_file_utils/directory_repository.py (hunk 3 above)
Comment (stephenfuqua) [reply src/file-utils/lms_file_utils/directory_repository.py:176]: done

File: utils/file-tester/lms_file_tester/__main__.py (hunk 5 above)
Comment (stephenfuqua) [reply utils/file-tester/lms_file_tester/__main__.py:134]: Done

**Review by lalrg — APPROVED** [review]
Looks good to me

**Pull request merged**